code/modules/xenoarcheaology/effects/cellcharge.dm

```diff
@@ -21,39 +41,49 @@
 	var/turf/turf = get_turf(holder)
 	if (!turf)
 		return
-	for (var/obj/machinery/power/apc/apc as anything in GLOB.apcs)
-		if (get_dist(turf, apc) <= 200)
-			for (var/obj/item/cell/cell in apc.contents)
-				cell.charge += 25
-	for (var/obj/machinery/power/smes/smes as anything in GLOB.smeses)
-		if (get_dist(turf, smes) <= effectrange)
-			smes.charge += 25
-	var/charged_robots
-	for (var/mob/living/silicon/robot/robot as anything in global.silicon_mob_list)
-		if (get_dist(turf, robot) < 50)
-			var/charged_cells
-			for (var/obj/item/cell/cell in robot.contents)
-				cell.charge += 25
-				charged_cells = TRUE
-			if (charged_cells)
-				to_chat(robot, "<font color='blue'>SYSTEM ALERT: Energy boost detected!</font>")
-				charged_robots = TRUE
-	if (charged_robots)
+
+	var/messaged_robots
+	for(var/mob/living/L in range(effectrange, turf))
+		var/obj/item/cell/C = L.get_cell()
+
+		if(C)
+			if(istype(L, /mob/living/silicon/robot) && ((last_message + (1 MINUTE)) < world.time))
+				messaged_robots = TRUE
+				to_chat(L, "<font color='blue'>SYSTEM ALERT: Energy boost detected!</font>")
+			C.charge = min(C.maxcharge, C.charge + 25)
+
+	if(messaged_robots)
 		last_message = world.time
 
+	for(var/obj/item/I in range(effectrange,turf))
+		var/obj/item/cell/C = I.get_cell()
+
+		if(C)
+			C.charge = min(C.maxcharge, C.charge + 25)
```

Comment:
All of this is copy-pasted thrice between touch, pulse, and aura, but for some numbers and a text string - can we squash this into a single proc?